doc/connectivity/networking/api/tls_credentials_shell.rst

```diff
@@ -0,0 +1,243 @@
+
+
+Arguments
+---------
+
+.. csv-table::
+   :header: "Argument", "Description"
+   :widths: 15 85
+
+   "``<SECTAG>``", "Optional. If provided, only list credentials with this sectag. Pass ``any`` or omit to allow any sectag. Otherwise, can be any non-negative integer."
+   "``<TYPE>``", "Optional. If provided, only list credentials with this credential type. Pass ``any`` or omit to allow any credential type. Otherwise, see :ref:`tls_credentials_shell_cred_types` for valid values."
+
+
+Output
+------
+
+The command outputs all matching credentials in the following (CSV-compliant) format:
+
+.. code-block:: shell
+
+   <SECTAG>,<TYPE>,<DIGEST>,<STATUS>
+
+Where:
+
+.. csv-table::
+   :header: "Symbol", "Value"
+   :widths: 15 85
+
+   "``<SECTAG>``", "The sectag of the listed credential. A non-negative integer."
+   "``<TYPE>``", "Credential type short-code (see :ref:`tls_credentials_shell_cred_types` for details) of the listed credential."
+   "``<DIGEST>``", "A string digest representing the credential contents. The exact nature of this digest may vary depending on credentials storage backend, but currently for all backends this is a base64 encoded SHA256 hash of the raw credential contents (so different storage formats for essentially identical credentials will have different digests)."
+   "``<STATUS>``", "Status code indicating success or failure with generating a digest of the listed credential. 0 if successful, negative error code specific to the storage backend otherwise. Lines for which status is not zero will be printed with error formatting."
+
+After the list is printed, a final summary of the found credentials will be printed in the form:
+
+.. code-block:: shell
+
+   <N> credentials found.
+
+Where `<N>` is the number of credentials found, and is zero if none are found.
+
+.. _tls_credentials_shell_cred_types:
+
+Credential Types
+****************
+
+The following keywords (case-insensitive) may be used to specify a credential type:
+
+.. csv-table::
+   :header: "Keyword(s)", "Meaning"
+   :widths: 15 85
+
+   "``CA_CERT``, ``CA``", "A trusted CA certificate."
+   "``SERVER_CERT``, ``SELF_CERT``, ``SELF``, ``SERV``", "Self or server certificate."
+   "``PRIVATE_KEY``, ``PK``", "A private key."
+   "``PRE_SHARED_KEY``, ``PSK``", "A pre-shared key."
+   "``PRE_SHARED_KEY_ID``, ``PSK_ID``", "ID for pre-shared key."
+
+.. _tls_credentials_shell_data_formats:
+
+Storage/Retrieval Formats
+*************************
+
+The following keywords (case-insensitive) specify how credential data is stored and retrieved.
+
+
+.. csv-table::
+   :header: "Keyword", "Meaning", "Behavior during storage (``cred add``)", "Behavior during retrieval (``cred get``)"
+   :widths: 3, 15, 41, 41
+
+   "``BIN``", "Credential is treated as non-terminated base64.", "Provided data will be decoded from base64 into raw binary before storage. No terminator will be appended.", "Stored data will be encoded into base64 before being printed."
```

Comment:
I find `non-terminated base64` to be confusing here.  There are two things going on which should be talked about separately:

1. how data is written and read via the shell: in base64, in both directions
2. how data is stored: in raw binary

Reply:
I agree, those two points are very important to make clear -- those are exactly what I was trying to communicate with the above sentences.

But maybe this is just a matter of phrasing -- would the following improve clarity in your opinion?

> Data read from the shell is interpreted as base64 and decoded into and stored as binary. No terminator will be appended.


I think that `Stored data will be encoded into base64 before being printed.` is already maximally clear, unless you have some specific suggestions


I could make similar changes for the other four formats

Reply:
The word `treated` could mean either the read/write or the storage; it's ambiguous.

Reply:
Oh good point, I had not thought of it in that way. `interpreted` is a much better word choice now that you mention it explicitly. Thank you. I'll switch to that.

Reply:
I misunderstood which `treated as` you were referring to. I will replace with `stored and retrieved as`

Reply:
I was referring to the use in the sentence: `Credential is treated as non-terminated base64.`  It's not treated the entire time as base64; if it were, it would be stored as base64.

Further, none of these options support raw binary as the input or output through the shell, nor reception of a NULL to terminate a string being input via the shell (input is terminated with a linefeed, correct?).

So really, the input/output format of any of your options is a string, and never anything else.  The string, depending on the name of the option, is either stored as-is, or decoded as base64 and stored that way, then encoded as base64 and displayed that way.

Reply:
> I was referring to the use in the sentence: Credential is treated as non-terminated base64

Gotcha, I'll be editing that too

> So really, the input/output format of any of your options is a string, and never anything else

Correct, what changes is what the string represents. The entered string either represents text (ie, string) data, or it represents binary data. So `bin/str` is supposed to denote what the user intends the entered string to represent

Perhaps `bin/txt` would be a little closer in meaning? I am really hesitant to use `b64` as the format specifier because I don't feel that that accurately captures that it is intended for providing essentially arbitrary binary data.

I worry it will make this already very difficult to describe situation more confusing to users